src/hera/workflows/_meta_mixins.py

```diff
@@ -275,17 +287,24 @@ def _get_param_items_from_source(source: Callable) -> List[Parameter]:
     List[Parameter]
         A list of identified parameters (possibly empty).
     """
-    source_signature: List[str] = []
-    for p in inspect.signature(source).parameters.values():
-        if p.default == inspect.Parameter.empty and p.kind == inspect.Parameter.POSITIONAL_OR_KEYWORD:
-            # only add positional or keyword arguments that are not set to a default value
-            # as the default value ones are captured by the automatically generated `Parameter` fields for positional
-            # kwargs. Otherwise, we assume that the user sets the value of the parameter via the `with_param` field
-            source_signature.append(p.name)
-
-    if len(source_signature) == 1:
-        return [Parameter(name=n, value="{{item}}") for n in source_signature]
-    return [Parameter(name=n, value=f"{{{{item.{n}}}}}") for n in source_signature]
+    non_default_parameters: List[Parameter] = []
+    if pydantic_input := _get_pydantic_input_type(source):
+        for parameter in pydantic_input._get_parameters():
+            if parameter.default is None:
+                non_default_parameters.append(parameter)
+    else:
+        for p in inspect.signature(source).parameters.values():
+            if p.default == inspect.Parameter.empty and p.kind == inspect.Parameter.POSITIONAL_OR_KEYWORD:
+                # only add positional or keyword arguments that are not set to a default value
+                # as the default value ones are captured by the automatically generated `Parameter` fields for positional
+                # kwargs. Otherwise, we assume that the user sets the value of the parameter via the `with_param` field
+                io = construct_io_from_annotation(p.name, p.annotation)
+                if isinstance(io, Parameter) and io.default is None and not io.output:
+                    non_default_parameters.append(io)
+
+    for param in non_default_parameters:
+        param.value = "{{" + ("item" if len(non_default_parameters) == 1 else f"item.{param.name}") + "}}"
```

Comment:
This seems harder to read/understand IMO

Reply:
It can't be as succinct as before because we're doing a side-effect. A couple of options spring to mind, which do you prefer?

```code
# Just move the if/else out into a separate line
for param in non_default_parameters:
    param_label = "item" if len(non_default_parameters) == 1 else f"item.{param.name}"
    param.value = "{{" + param_label + "}}"
```

```code
# Move the if/else back out to the top
if len(non_default_parameters) == 1:
    non_default_parameters[0].value = "{{item}}"
else:
    for param in non_default_parameters:
        param.value = "{{{item." + param.name + "))
```